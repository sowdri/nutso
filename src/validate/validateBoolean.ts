import { BooleanResult } from "../models/result/BooleanResult";
import { BooleanSchema } from "../models/schema/BooleanSchema";
import { optionalFlagValidator } from "../utils/optionalFlagValidator";
import { isBoolean, isNil } from "../utils/typeChecker";
import { validationFnExecutor } from "../utils/validationFnExecutor";

export const validateBoolean = <P>(args: { value: any; parent: P; schema: BooleanSchema<P> }): BooleanResult => {
  const { value, parent, schema } = args;
  // isnil
  if (isNil(value)) {
    return optionalFlagValidator({ parent, flag: schema.optional });
  }

  if (!isBoolean(value)) {
    return {
      isValid: false,
      errorMessage: `Should be true or false.`,
      errorPath: [],
    };
  }

  // validationFn
  if (schema.validationFn) {
    const result = validationFnExecutor({ value, validationFn: schema.validationFn, parent });
    if (result) return result;
  }

  return {
    isValid: true,
    errorMessage: ``,
    errorPath: [],
  };
};
