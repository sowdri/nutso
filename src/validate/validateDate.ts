import { DateResult } from "../models/result/DateResult";
import { DateSchema } from "../models/schema/DateSchema";
import { isDate, isNil } from "../utils/typeChecker";
import { validationFnExecutor } from "../utils/validationFnExecutor";
import { optionalFlagValidator } from "../utils/optionalFlagValidator";

export const validateDate = <R, P>(args: { value: any; root: R; parent: P; schema: DateSchema<R, P> }): DateResult => {
  const { value, schema } = args;
  // isnil
  if (isNil(value)) {
    return optionalFlagValidator({ ...args, flag: schema.optional });
  }

  // is date object
  if (!isDate(value)) {
    return {
      isValid: false,
      errorMessage: `Should be a valid date.`,
      errorPath: [],
    };
  }

  const date = value;

  // custom validationFn
  if (schema.validationFn) {
    const result = validationFnExecutor({ ...args, value: date, validationFn: schema.validationFn });
    if (result) return result;
  }

  return {
    isValid: true,
    errorMessage: ``,
    errorPath: [],
  };
};
