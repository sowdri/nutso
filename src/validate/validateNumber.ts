import { NumberResult } from "../models/result/NumberResult";
import { NumberSchema } from "../models/schema/NumberSchema";
import { optionalFlagValidator } from "../utils/optionalFlagValidator";
import { isNil, isNumber } from "../utils/typeChecker";
import { validationFnExecutor } from "../utils/validationFnExecutor";

export const validateNumber = <R, P>(args: {
  value: any;
  root: R;
  parent: P;
  schema: NumberSchema<R, P>;
}): NumberResult => {
  //
  const { value, schema } = args;
  // isnil
  if (isNil(value)) {
    return optionalFlagValidator({ ...args, flag: schema.optional });
  }

  if (!isNumber(value) || isNaN(value)) {
    return {
      isValid: false,
      errorMessage: `Should be a number.`,
      errorPath: [],
    };
  }

  const numbr = value as number;

  // min length
  if (!isNil(schema.min) && numbr < schema.min!) {
    return {
      isValid: false,
      errorMessage: `Should not be less than ${schema.min}.`,
      errorPath: [],
    };
  }

  // max length
  if (!isNil(schema.max) && numbr > schema.max!) {
    return {
      isValid: false,
      errorMessage: `Should not be larger than ${schema.max}.`,
      errorPath: [],
    };
  }

  // pattern
  if (!isNil(schema.pattern)) {
    const match = schema.pattern!.test(value);
    if (!match) {
      return {
        isValid: false,
        errorMessage: `Should match the pattern ${schema.pattern} .`,
        errorPath: [],
      };
    }
  }

  // validationFn
  if (schema.validationFn) {
    const result = validationFnExecutor({ ...args, value: numbr, validationFn: schema.validationFn });
    if (result) return result;
  }

  return {
    isValid: true,
    errorMessage: ``,
    errorPath: [],
  };
};
