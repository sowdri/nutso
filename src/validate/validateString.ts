import { StringResult } from "../models/result/StringResult";
import { StringSchema } from "../models/schema/StringSchema";
import { isNil, isString } from "../utils/typeChecker";
import { validationFnExecutor } from "../utils/validationFnExecutor";
import { optionalFlagValidator, isOptional } from "../utils/optionalFlagValidator";

export const validateString = <R, P>(args: {
  value: any;
  root: R;
  parent: P;
  schema: StringSchema<R, P>;
}): StringResult => {
  const { value, schema } = args;
  //

  // isnil
  if (isNil(value)) {
    return optionalFlagValidator({ ...args, flag: schema.optional });
  }

  // check if type is string
  if (!isString(value)) {
    return {
      isValid: false,
      errorMessage: `Should be a string.`,
      errorPath: [],
    };
  }

  const str = value as string;

  // check if empty
  if (str === "") {
    const optional = isOptional({ ...args, flag: schema.optional });
    if (!optional)
      return {
        isValid: false,
        errorMessage: `Should not be empty.`,
        errorPath: [],
      };
  }

  // min length
  if (!isNil(schema.minLength) && str.length < schema.minLength!) {
    return {
      isValid: false,
      errorMessage: `Should be at least ${schema.minLength} characters.`,
      errorPath: [],
    };
  }

  // max length
  if (!isNil(schema.maxLength) && str.length > schema.maxLength!) {
    return {
      isValid: false,
      errorMessage: `Should not be longer than ${schema.maxLength} characters.`,
      errorPath: [],
    };
  }

  // pattern
  if (!isNil(schema.pattern)) {
    const match = schema.pattern!.test(value);
    if (!match) {
      return {
        isValid: false,
        errorMessage: `Should match the pattern ${schema.pattern} .`,
        errorPath: [],
      };
    }
  }

  // validationFn
  if (schema.validationFn) {
    const result = validationFnExecutor({ ...args, value: str, validationFn: schema.validationFn });
    if (result) return result;
  }

  return {
    isValid: true,
    errorMessage: ``,
    errorPath: [],
  };
};
