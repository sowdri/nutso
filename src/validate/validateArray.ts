import { ArrayResult } from "../models/result/ArrayResult";
import { ArraySchema } from "../models/schema/ArraySchema";
import { optionalFlagValidator } from "../utils/optionalFlagValidator";
import { isNil } from "../utils/typeChecker";
import { _validate } from "./validate";

export const validateArray = <E, T extends E[], R, P>(args: {
  value: T | null;
  root: R;
  parent: P;
  schema: ArraySchema<E, T, R, P>;
}): ArrayResult<E> => {
  const { value: arr, schema } = args;
  // isnil
  if (isNil(arr)) {
    return { ...optionalFlagValidator({ ...args, flag: schema.optional }), items: [] };
  }

  const result: ArrayResult<E> = {
    isValid: true,
    errorMessage: "",
    items: [],
    errorPath: [],
  };

  // array min-items
  if (!isNil(schema.minItems) && arr.length < schema.minItems!) {
    result.isValid = false;
    result.errorMessage = `Should have at least ${schema.minItems} items.`;
  }

  // array max-items
  if (!isNil(schema.maxItems) && arr.length >= schema.maxItems!) {
    result.isValid = false;
    result.errorMessage = `Should not have more than ${schema.maxItems! - 1} items.`;
  }

  // for each key, validate
  for (let i = 0; i < arr.length; i++) {
    result.items[i] = _validate({ ...args, value: arr[i], parent: arr as any, schema: schema.items });
  }

  // if this node is valid, then check if all of it's children are valid
  // because the node is invalid, if any of it's children are invalid
  if (result.isValid) {
    for (let i = 0; i < arr.length; i++) {
      const item = result.items[i];
      if (!item.isValid) {
        result.isValid = false;
        result.errorMessage = item.errorMessage;
        result.errorPath = [i, ...item.errorPath];
        break;
      }
    }
  }

  return result;
};
