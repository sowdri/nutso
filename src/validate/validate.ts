import { Result } from "../models/result/Result";
import { BooleanSchema } from "../models/schema/BooleanSchema";
import { DateSchema } from "../models/schema/DateSchema";
import { NumberSchema } from "../models/schema/NumberSchema";
import { Schema } from "../models/schema/Schema";
import { StringSchema } from "../models/schema/StringSchema";
import { validateArray } from "./validateArray";
import { validateBoolean } from "./validateBoolean";
import { validateDate } from "./validateDate";
import { validateNumber } from "./validateNumber";
import { validateObject } from "./validateObject";
import { validateString } from "./validateString";

/**
 * This is an internal API that takes the root as an argument
 * and called recursively from `validateObject` and `validateArray` methods
 *
 * @param value
 * @param parent
 * @param schema
 */
export const _validate = <T, P>(args: { value: T | null; parent: P; schema: Schema<T, P> }): Result<T> => {
  //

  if (!args.schema) {
    throw new Error("Schema should not be null");
  }

  /**
   * validate object against the schema
   */
  switch (args.schema.type) {
    case "number":
      return validateNumber({ ...args, schema: args.schema as NumberSchema<P> }) as Result<T>;
    case "string":
      return validateString({ ...args, schema: args.schema as StringSchema<P> }) as Result<T>;
    case "boolean":
      return validateBoolean({ ...args, schema: args.schema as BooleanSchema<P> }) as Result<T>;
    case "array":
      return validateArray({ value: args.value as any, parent: args.parent, schema: args.schema as any }) as Result<T>;
    case "object":
      return validateObject({ ...args, schema: args.schema }) as Result<T>;
    case "date":
      return validateDate({ ...args, schema: args.schema as DateSchema<P> }) as Result<T>;
  }
  throw new Error(`Unhandled data type`);
};

// public api
export const validate = <T>(value: T, schema: Schema<T>): Result<T> => {
  return _validate({ value, parent: null as any, schema });
};
