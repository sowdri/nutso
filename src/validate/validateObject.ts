import { ObjectResult } from "../models/result/ObjectResult";
import { ObjectSchema } from "../models/schema/ObjectSchema";
import { optionalFlagValidator } from "../utils/optionalFlagValidator";
import { isNil } from "../utils/typeChecker";
import { _validate } from "./validate";

export const isRegex = (str: string) => {
  return str.startsWith("/") && str.endsWith("/");
};

export const getRegex = (str: string) => {
  return new RegExp(`^${str.slice(1, -1)}$`);
};

export const validateObject = <T extends { [key: string]: any }, R, P>(args: {
  value: T | null;
  root: R;
  parent: P;
  schema: ObjectSchema<T, R, P>;
}): ObjectResult<T> => {
  //
  const { value, schema } = args;

  const result: ObjectResult<T> = {
    isValid: true,
    errorMessage: "",
    properties: {} as any,
    errorPath: [],
  };

  // isnil
  if (isNil(value)) {
    return { ...optionalFlagValidator({ ...args, flag: schema.optional }), properties: {} as any };
  }

  const obj = value as T;
  const processedFields: string[] = [];

  // for each static key, validate
  for (let field in schema.properties) {
    if (isRegex(field)) continue;
    result.properties[field] = _validate({
      ...args,
      value: value ? value[field] : null,
      parent: value,
      schema: schema.properties[field],
    });
    processedFields.push(field);
  }

  // for each regex validate all matching keys
  for (let field in schema.properties) {
    if (!isRegex(field)) continue; // not-regex
    const regex = getRegex(field);
    // for each key
    for (let key in obj) {
      /**
       * Already processed field. This will happen if the regex matches a wide range of fields, like ".*"
       * and the properties also defines specific fields like "red": {}, "blue": {} etc ...
       */
      if (processedFields.includes(key)) continue;
      if (!regex.test(key)) continue;
      result.properties[key] = _validate({
        ...args,
        value: value ? value[key] : null,
        parent: value,
        schema: schema.properties[field],
      });
      processedFields.push(key);
    }
  }

  // if this node is valid, then check if all of it's children are valid
  // because the node is invalid, if any of it's children are invalid
  if (result.isValid) {
    for (let field of processedFields) {
      const property = result.properties[field];
      if (!property.isValid) {
        result.isValid = false;
        result.errorMessage = property.errorMessage;
        result.errorPath = [field, ...property.errorPath];
        break;
      }
    }
  }

  return result;
};
