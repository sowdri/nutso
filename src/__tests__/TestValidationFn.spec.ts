import { Schema } from "..";
import { validate } from "../validate/validate";

test(`Compose schema with different roots - typesafety`, () => {
  type Address = {
    line1: string;
  };
  type Customer = {
    name: string;
    address: Address;
  };

  const addressSchema: Schema<Address, Customer> = {
    type: "object",
    properties: {
      line1: {
        type: "string",
        validationFn: (args) => {
          args.parent.line1;
          return {
            isValid: true,
            errorMessage: "",
          };
        },
      },
    },
  };

  const customerSchema: Schema<Customer> = {
    type: "object",
    properties: {
      name: {
        type: "string",
        validationFn: (args) => {
          return {
            isValid: false,
            errorMessage: "Custom validation failed",
          };
        },
      },
      address: addressSchema,
    },
  };

  // this test is just to ensure typescript is not complaining!
  // so no assertion required
  expect(true).toBeTruthy();
});

test(`Custom validation - check errorPath - level 1`, () => {
  type Address = {
    line1: string;
  };
  type Customer = {
    name: string;
    address: Address;
  };

  const addressSchema: Schema<Address, Customer> = {
    type: "object",
    properties: {
      line1: {
        type: "string",
        validationFn: (args) => {
          args.parent.line1;
        },
      },
    },
  };

  const customerSchema: Schema<Customer> = {
    type: "object",
    properties: {
      name: {
        type: "string",
        validationFn: (args) => {
          return {
            isValid: false,
            errorMessage: "Custom validation failed",
          };
        },
      },
      address: addressSchema,
    },
  };

  const customer: Customer = {
    name: "John",
    address: {
      line1: "10 Downing St",
    },
  };

  const result = validate(customer, customerSchema);
  expect(result).toMatchSnapshot();
});

test(`Custom validation - check errorPath - level 2`, () => {
  type Address = {
    line1: string;
  };
  type Customer = {
    name: string;
    address: Address;
  };

  const addressSchema: Schema<Address, Customer> = {
    type: "object",
    properties: {
      line1: {
        type: "string",
        validationFn: (args) => {
          return {
            errorMessage: "Custom validation failed",
          };
        },
      },
    },
  };

  const customerSchema: Schema<Customer> = {
    type: "object",
    properties: {
      name: {
        type: "string",
        validationFn: (args) => {},
      },
      address: addressSchema,
    },
  };

  const customer: Customer = {
    name: "John",
    address: {
      line1: "10 Downing St",
    },
  };

  const result = validate(customer, customerSchema);
  expect(result).toMatchSnapshot();
});

test(`Password equality validation`, () => {
  type LoginForm = {
    username: string;
    password: string;
    repeatPassword: string;
  };

  const loginSchema: Schema<LoginForm> = {
    type: "object",
    properties: {
      username: {
        type: "string",
      },
      password: {
        type: "string",
        pattern: /\w{6}/,
      },
      repeatPassword: {
        type: "string",
        validationFn: (args) => {
          if (args.value !== args.parent.password)
            return {
              errorMessage: "Passwords do not match",
            };
        },
      },
    },
  };

  const form1: LoginForm = {
    username: "John",
    password: "foobarbaz",
    repeatPassword: "foofoofoo",
  };
  const result = validate(form1, loginSchema);
  expect(result).toMatchSnapshot();
});
