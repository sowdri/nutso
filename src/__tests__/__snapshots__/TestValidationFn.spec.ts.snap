// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom validation - check errorPath - level 1 1`] = `
{
  "errorMessage": "Custom validation failed",
  "errorPath": [
    "name",
  ],
  "isValid": false,
  "properties": {
    "address": {
      "errorMessage": "",
      "errorPath": [],
      "isValid": true,
      "properties": {
        "line1": {
          "errorMessage": "",
          "errorPath": [],
          "isValid": true,
        },
      },
    },
    "name": {
      "errorMessage": "Custom validation failed",
      "errorPath": [],
      "isValid": false,
    },
  },
}
`;

exports[`Custom validation - check errorPath - level 2 1`] = `
{
  "errorMessage": "Custom validation failed",
  "errorPath": [
    "address",
    "line1",
  ],
  "isValid": false,
  "properties": {
    "address": {
      "errorMessage": "Custom validation failed",
      "errorPath": [
        "line1",
      ],
      "isValid": false,
      "properties": {
        "line1": {
          "errorMessage": "Custom validation failed",
          "errorPath": [],
          "isValid": false,
        },
      },
    },
    "name": {
      "errorMessage": "",
      "errorPath": [],
      "isValid": true,
    },
  },
}
`;

exports[`Password equality validation 1`] = `
{
  "errorMessage": "Passwords do not match",
  "errorPath": [
    "repeatPassword",
  ],
  "isValid": false,
  "properties": {
    "password": {
      "errorMessage": "",
      "errorPath": [],
      "isValid": true,
    },
    "repeatPassword": {
      "errorMessage": "Passwords do not match",
      "errorPath": [],
      "isValid": false,
    },
    "username": {
      "errorMessage": "",
      "errorPath": [],
      "isValid": true,
    },
  },
}
`;
